### ISSUE TITLE
[Feature Request] Sparse compute_gradient

### ISSUE THREAD

**Issue by OscarDPan** [tensorflow/tensorflow#15464]
I am working on an extremely large scale linear model and have been trying to optimize the performance of the TF optimizer.

**Have I written custom code**
Version I.
My feature size is huge (500Mil) and sparse, so I was testing if I could make TF only compute the necessary gradients and apply it using some function like tf.scatter_sub()
My graph is like this:
```
cost = fn(w)
vars_to_update = tf.gather(w, non_zero_indices)
grads = tf.gradients(cost, vars_to_update)
update_op = tf.scatter_sub(w, non_zero_indeces, grads)
```
I found that tf.graidents() always returns None for tf.gather(). Similar condition if I pass tf.gather() to any optimizer like tf.train.GradientDescentOpitmizer(cost, tf.gather(w, indices)), it will throw unsupported error for tf.gather(). 

I was wondering if I did anything wrong or TF just doesn't support sparse gradient computation? If latter does TF team plan to have that implemented in short future?

Version II.
In stead of creating a sparse tensor and do sparse_tensor_dense_matmul(), I also tried using tf.gather() follow by tf.segment_sum() to implement W*X. By doing this the optimizer apparently automatically performed sparse grad computation and sparse update. However, the speed of the optimizer was **horribly slower (15seconds)** than the sparse tensor approach. And idea why?

Pseudo code:
```
active_weights = tf.gather(weights, non_zero_indices)
total = tf.segment_sum(
                tf.reshape(activated_weights, [-1]),
                segment_ids //which is the row number e.g. [0,0,0,0,1,1,1,2,2,2,3,3,...]
                )
update_op = tf.train.GradientDescentOptimizer().minimize(total, active_weights)
```

**OS Platform and Distribution**
Centos Linux version 3.10.0-229.4.2.el7.x86_64 (gcc version 4.8.2 20140120 (Red Hat 4.8.2-16) (GCC) )

**TensorFlow installed from (source or binary)**
pip install tensorflow

**TF version**
1.3.0

**Python version**
2.7.5

**Bazel version, CUDA/cuDNN, GPU model and memory, Exact command to reproduce**
N/A

**Also there might be a potential bug**
If in the second approach (tf.gather() and tf.segment_sum()) I replace GradientDescentOptimizer with Adam or Adagrad optimizer, the memory would blow up very quickly. I did not look into why that happened so I am not sure if this worth a bug ticket.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by OscarDPan** [comment]
@tensorflowbutler Thanks for the reminder. I've updated my post according to your suggestions.

**Comment by formath** [comment]
Did you try `embedding_lookup_sparse`?

**Comment by OscarDPan** [comment]
@formath Thanks for your suggestion. I am not sure how embddeing_lookup_sparse help in my case. Could you elaborate?

**Comment by formath** [comment]
```
def _do_gather(params, ids, name=None):
  """Deals with doing gather differently for resource variables."""
  if isinstance(params, resource_variable_ops.ResourceVariable):
    return params.sparse_read(ids, name=name)
  return array_ops.gather(params, ids, name=name)
```
`embedding_lookup_sparse` uses `sparse_read` so may be more efficient.  Before you try, I'm not sure about how much improvement you will get or even whether it will works.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by drpngx** [comment]
@OscarDPan how did that work out for you?

**Comment by OscarDPan** [comment]
@drpngx Yes I have found the solution. The trick is to use tf.scatter_sub().op
